clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2682,19 +2704,53 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
   }
   O << "};\n\n";
 
+  O << "// array representing kernel parameters used in all kernels defined in "
+       "the\n";
+  O << "// corresponding source\n";
+  O << "static constexpr\n";
+  O << "const bool kernel_param_used[] = {\n";
+  O << "  //PARAM_USED_TABLE_BEGIN \n";
+  for (unsigned I = 0; I < KernelDescs.size(); I++) {
+    auto &K = KernelDescs[I];
+    O << "  //--- " << K.Name << "\n";
+    O << "  ";
+    for (unsigned I = 0; I < K.Params.size(); I++) {
+      auto &P = K.Params[I];
+      for (unsigned J = 0; J < P.Count; J++) {
+        O << "1";
+        O << ", ";
+      }
+    }
+    O << "\n";
+  }
+  O << "  //PARAM_USED_TABLE_END "
+    << "\n";
+  O << "  } \n\n";
+
   O << "// array representing signatures of all kernels defined in the\n";
   O << "// corresponding source\n";
   O << "static constexpr\n";
   O << "const kernel_param_desc_t kernel_signatures[] = {\n";
 
+  unsigned CurIndex = 0;
   for (unsigned I = 0; I < KernelDescs.size(); I++) {
     auto &K = KernelDescs[I];
     O << "  //--- " << K.Name << "\n";
-
-    for (const auto &P : K.Params) {
+    for (unsigned I = 0; I < K.Params.size(); I++) {
+      auto &P = K.Params[I];
       std::string TyStr = paramKind2Str(P.Kind);
       O << "  { kernel_param_kind_t::" << TyStr << ", ";
-      O << P.Info << ", " << P.Offset << " },\n";
+      O << P.Info << ", " << P.Offset << ", ";
+      for (unsigned X = 0; X < P.Count; X++) {
+        O << "("
+          << "kernel_param_used[" << CurIndex << "]"
+          << " << " << X << ")";
+        if (X < P.Count - 1)
+          O << " | ";
+        CurIndex = CurIndex + 1;
+      }
+      O << "}";
+      O << ",\n";
```

Comment:
Is the extra "," ok here? Other places had comma even after the last element in the  array.